lib/kubernetes-deploy/kubernetes_resource.rb

```diff
@@ -54,19 +41,28 @@ def timeout
       self.class.timeout
     end
 
-    def initialize(name:, namespace:, context:, file:, logger:)
+    def initialize(namespace:, context:, template:, logger:)
       # subclasses must also set these if they define their own initializer
-      @name = name
+      @name = template.fetch("metadata", {})["name"]
       @namespace = namespace
       @context = context
-      @file = file
       @logger = logger
+      @template = template
     end
 
     def id
       "#{type}/#{name}"
     end
 
+    def file
```

Comment:
> This can also result to tempfiles not being garbage collected after the process is done...

What can result in that? Tempfiles are automatically GC'd when the process exits.

> IMO when a method returns an instance of File (or any IO) you'd expect to read it directly without the need for File.read(file.path). Though it's only used for debugging and it's not critical.

Hm. The thing is that for kubectl to be able to read it, it needs to be closed. It seems to me having something external close it would be more confusing overall. Idea: We could make this method private, and only expose one called `file_path`. Then callers clearly need to do `File.read`. WDYT?